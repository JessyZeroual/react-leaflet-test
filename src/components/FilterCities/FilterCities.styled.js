import styled from "styled-components";
import { secondaryColor } from "../../constant/style-constants";
import { StyledWrapperCenter } from "../../style/common.styled-component";

export const StyledWrapperCityNotAvailable = styled(StyledWrapperCenter)`
  padding: 10px;
  color: #d9534f;
`;

export const StyledInputSearch = styled.input`
  outline: none;
  border: none;
  margin-left: 10px;
  height: 30px;
  width: 100%;
`;

export const StyledInputGroup = styled.div`
  display: flex;
  height: 30px;
`;

export const StyledForm = styled.form`
  padding: 2px 10px;
  background: #fff;
  border-radius: 20px;
  overflow: hidden;
  &:hover {
    box-shadow: 0 0 3pt grey;
  }
`;

export const StyledWrapperSearchResult = styled.div`
  margin-top: 5px;
  border-top: 1px solid #c0c0c0;
`;

export const StyledItemSearchResult = styled.div`
  padding: 10px;
  color: ${({ active }) => active && secondaryColor};
  &:hover {
    cursor: pointer;
    color: ${secondaryColor};
  }
`;
