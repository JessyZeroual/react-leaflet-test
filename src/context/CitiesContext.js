import React from "react";

const CitiesContext = React.createContext({});

export const CitiesContextProvider = CitiesContext.Provider;
export const CitiesContextConsumer = CitiesContext.Consumer;
export default CitiesContext;
