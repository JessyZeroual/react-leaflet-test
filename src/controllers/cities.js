export const getCities = () => {
  return fetch(`/cities.json`)
    .then((res) => res.json())
    .then((data) =>
      data.map((city) => ({
        ...city,
        growth_from_2000_to_2013: Number(
          city.growth_from_2000_to_2013.replace("%", "")
        ),
        population: Number(city.population),
        rank: Number(city.rank),
      }))
    );
};
