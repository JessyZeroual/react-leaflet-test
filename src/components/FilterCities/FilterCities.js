import React, { useState, useContext } from "react";
import CitiesContext from "../../context/CitiesContext";
import {
  StyledInputSearch,
  StyledInputGroup,
  StyledForm,
  StyledWrapperSearchResult,
  StyledItemSearchResult,
  StyledWrapperCityNotAvailable,
} from "./FilterCities.styled";

const FilterCities = () => {
  const citiesContext = useContext(CitiesContext);
  const { cities, currentCity } = citiesContext.state;
  const { setCurrentCity } = citiesContext.dispatch;
  const [searchPhrase, setSearchPhrase] = useState("");

  const citySuggestions = cities.filter(({ city }) =>
    city.toLowerCase().includes(searchPhrase.toLowerCase())
  );

  return (
    <StyledForm onSubmit={(e) => e.preventDefault()}>
      <StyledInputGroup>
        <img src="/icon/search.svg" width="15" alt="icon search" />
        <StyledInputSearch onChange={(e) => setSearchPhrase(e.target.value)} />
      </StyledInputGroup>

      {searchPhrase && (
        <StyledWrapperSearchResult>
          {citySuggestions.length ? (
            citySuggestions.slice(0, 5).map((citySuggestion, key) => (
              <StyledItemSearchResult
                key={key}
                active={currentCity.city === citySuggestion.city}
                onClick={() => setCurrentCity(citySuggestion)}
              >
                {citySuggestion.city}
              </StyledItemSearchResult>
            ))
          ) : (
            <StyledWrapperCityNotAvailable>
              This city is not available
            </StyledWrapperCityNotAvailable>
          )}
        </StyledWrapperSearchResult>
      )}
    </StyledForm>
  );
};

export default FilterCities;
