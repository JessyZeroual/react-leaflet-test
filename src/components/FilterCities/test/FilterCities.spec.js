import React from "react";
import TestRenderer from "react-test-renderer";
import { act } from "react-test-renderer";

import FilterCities from "../FilterCities";
import CitiesContext from "../../../context/CitiesContext";
import cities from "../../../test/listOfCitiesTests";

const currentCity = [
  {
    city: "New York",
    growth_from_2000_to_2013: "4.8%",
    latitude: 40.7127837,
    longitude: -74.0059413,
    population: "8405837",
    rank: "1",
    state: "New York",
  },
];

const setCurrentCity = jest.fn();

describe("FilterCities", () => {
  let renderer;

  const mountFilterCities = () =>
    act(() => {
      renderer = TestRenderer.create(
        <CitiesContext.Provider
          value={{
            state: { cities, currentCity },
            dispatch: { setCurrentCity },
          }}
        >
          <FilterCities />
        </CitiesContext.Provider>
      );
    });

  describe("When FilterCities is mounted", () => {
    it("renders correctly", () => {
      mountFilterCities();
      expect(renderer).toMatchSnapshot();
    });

    describe("When we handle input", () => {
      mountFilterCities();
      describe("When you enter an existing value in the list of cities", () => {
        it("returns the cities containing the entered value", async () => {
          const input = renderer.root.findByType("input");

          await act(async () => {
            input.props.onChange({ target: { value: "n" } });
          });

          expect(renderer.toJSON().children[1].children[0].children).toEqual([
            "New York",
          ]);
        });
      });

      describe("When you enter a value that does not exist in the list of cities", () => {
        it("returns 'This city is not available' ", async () => {
          const input = renderer.root.findByType("input");

          await act(async () => {
            input.props.onChange({ target: { value: "FakeCity" } });
          });

          expect(renderer.toJSON().children[1].children[0].children).toEqual([
            "This city is not available",
          ]);
        });
      });
    });
  });
});
