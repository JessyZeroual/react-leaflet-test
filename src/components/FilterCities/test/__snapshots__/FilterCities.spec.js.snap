// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterCities When FilterCities is mounted renders correctly 1`] = `
<form
  className="sc-AxhUy eElOBO"
  onSubmit={[Function]}
>
  <div
    className="sc-AxhCb bsfCal"
  >
    <img
      alt="icon search"
      src="/icon/search.svg"
      width="15"
    />
    <input
      className="sc-AxiKw VVkvQ"
      onChange={[Function]}
    />
  </div>
  
</form>
`;
